cls_module/cls_module/memory/stm/aha.py

```diff
@@ -315,23 +402,60 @@ def forward_memory(self, inputs, targets, labels):
 
     losses = {}
     outputs = {}
+    features = {}
 
     outputs['ps'] = self.forward_ps(inputs)
+    features['ps'] = outputs['ps'].detach().cpu()
 
     # Compute DG Overlap
     overlap = self.ps.compute_overlap(outputs['ps'])
     losses['ps_overlap'] = overlap.sum()
 
-    pr_targets = outputs['ps'] if self.training else self.pc_buffer_batch
-    losses['pr'], outputs['pr'] = self.forward_pr(inputs=inputs, targets=pr_targets)
+    # Perforant Pathway: Hebbian Learning
+    if self.is_hebbian_perforant():
+      with torch.no_grad():
+        dg_ca3_in = outputs['ps']
+        pre_dg_ca3_out = self.dg_ca3(dg_ca3_in)
 
-    # Compute PR Mismatch
-    pr_out = outputs['pr']['pr_out']
-    pr_batch_size = pr_out.shape[0]
-    losses['pr_mismatch'] = torch.sum(torch.abs(pr_targets - pr_out)) / pr_batch_size
+        ec_ca3_in = torch.flatten(inputs, 1)
+        pre_ec_ca3_out = self.ec_ca3(ec_ca3_in)
 
-    pc_cue = outputs['ps'] if self.training else outputs['pr']['z_cue']
-    outputs['pc'] = self.forward_pc(inputs=pc_cue)
+        pre_pc_out = pre_dg_ca3_out + pre_ec_ca3_out
+        outputs['pc'] = pre_pc_out
+
+        if self.training:
+          # Update DG:CA3 with respect to dg_ca3_in (i.e. outputs['ps'])
+          d_dg_ca3 = self.learning_rule.update(dg_ca3_in, pre_pc_out, self.dg_ca3.weight)
+          d_dg_ca3 = d_dg_ca3.view(*self.dg_ca3.weight.size())
+          self.dg_ca3_optimizer.local_step(d_dg_ca3)
+
+          # Update EC:CA3 with respect to ec_ca3_in (i.e. inputs)
+          d_ec_ca3 = self.learning_rule.update(ec_ca3_in, pre_pc_out, self.ec_ca3.weight)
+          d_ec_ca3 = d_ec_ca3.view(*self.ec_ca3.weight.size())
+          self.ec_ca3_optimizer.local_step(d_ec_ca3)
+
+        post_dg_ca3_out = self.dg_ca3(dg_ca3_in)
```

Comment:
Just confirming - is it just to compute loss?

Reply:
I added a comment to clarify this. We only run it again to compute the loss between pre and post update activity